File: 491-Increasing Subsequences-递增子序列/491-Increasing Subsequences-递增子序列.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
 let res = []
 let path = []
 var findSubsequences = function(nums) {
     res = []
     backtrace(nums, 0)
     return res
 };
 
 var backtrace = function(nums, index){
     if(index > nums.length) return
 
     if(path.length >= 2) res.push(Array.from(path))
     
     let used = []
 
     for(let i = index; i < nums.length; i++){
         // 同层去重
         if(i > index && used[nums[i]]) continue
         used[nums[i]] = true
         // 排除非递增序列
         if(nums[i] < nums[index - 1]) continue
 
         path.push(nums[i])
         backtrace(nums, i + 1)
         path.pop()
     }
 }

File: 491-Increasing Subsequences-递增子序列/491-Increasing Subsequences-递增子序列.java
class Solution {
    private List<List<Integer>> res;
    private LinkedList<Integer> path = new LinkedList<Integer>();
    public List<List<Integer>> findSubsequences(int[] nums) {
        res = new ArrayList<>();
        backtrace(nums, 0);
        return res;
    }

    public void backtrace(int[] nums, int index){
        if(index > nums.length) return;

        if(path.size() >= 2){
            res.add(new ArrayList<Integer>(){{addAll(path);}});
        }

        boolean[] used = new boolean[202];
        for(int i = index; i < nums.length; i++){
            if(index < i && used[nums[i] + 100]) continue;
            used[nums[i] + 100] = true;
            if(path.size() != 0 && nums[i] < path.get(path.size() - 1) ) continue;

            path.add(nums[i]);
            backtrace(nums, i + 1);
            path.removeLast();
        }
    }
}